Add unpackaged support to PushNotifications spec (#1064)
microsoft/WindowsAppSDK: specs/PushNotifications/PushNotifications-spec.md

    Defines the classes that encapsulate Push notifications. (UWP and Win32)
-   [The Windows Push Notification Services (WNS) Overview](https://docs.microsoft.com/windows/uwp/design/shell/tiles-and-notifications/windows-push-notification-services--wns--overview)
    Defines how an app today can subscribe to Push Notifications and send a payload down to the
    corresponding client.

The below diagram is an illustration of the overall Push Flow that we have today:
![Legacy Flow For Windows App SDK](Legacy.png)

## The problems today

**Portal Registration**: The entire Registration flow for Push is dependent on registering an app in
the partner center portal by
[associating an app with the Windows Store](https://docs.microsoft.com/azure/notification-hubs/notification-hubs-windows-store-dotnet-get-started-wns-push-notification).
As a result, non-store provisioned apps cannot use this feature.

**Client Registration**: Unpackaged 3rd party applications cannot register themselves as a Push
target. The only way to get an implicit client side registration is to package the app as an MSIX

> Desktop Bridge or SSP which may not satisfy 3rd party requirements.

Author: SSP = Sparse signed packages

**Push Channels**: All the channel APIs are dependent on the caller app having a store provisioned
identity. This means that Win32 and unpackaged apps that are not store distributed cannot use our
Public APIs. Sideloaded UWPs have exactly the same problem.

**Activation**: When the OS receives a Push over the WNS socket, it signals the Background
Infrastructure to trigger a task and run some code on the app's behalf. This flow is undefined for
Win32 apps that need to use Push today due to the limitations described above.

**Channel Request Error Handling**: The channel APIs exposed today can throw exceptions via their
Projections Some errors can be critical errors while others are retryable errors. The msdn guidance
for retrying isn't very clear and has been the root cause of multiple app bugs in the past. We want
to abstract away the details of channel retry operations for developers in Windows App SDK.

# Description

At a high level, we need to provide a way for all Win32 applications to subscribe to and receive
Push notifications irrespective of their app type. This includes unpackaged apps and packaged win32

> (MSIX Desktop Bridge, Sparse Signed Packages). Moreover, all Push scenarios should adhere to OS

Reviewer: You let out Win32App

...(MSIX Desktop Bridge, MSIX Win32App, Sparse Signed Packages)...

Win32App aka RuntimeBehavior=Win32App is like Desktop Bridge except instead of virtualized I/O and the like goodness, Win32App processes strongly favor (classic) Win32 compability. No Helium container, Loader has the classic 18+ mechanisms in the DLL Search Order vs 8, among other things.

P.S. Sparse Signed Packages isn't really another 'app type', more like a way of unpackaged apps opting in to MSIX goodness for runtime without replacing their installer with MSIX's. But sure, can't hurt to mention it here. All good.

> resource management policies like Power Saver, Network Attribution, Enterprise group policies etc.

Author: Network Attribution is how much data an app uses. I can add this as a simple definition in parenthesis.

To see this value: goto Settings->Apps->Apps&Features->Your App.

The Windows App SDK Push component will abstract away the complexities of dealing with WNS channel
registrations and Push related activations as much as possible freeing the developer to focus on
other app related challenges.

We will prioritize the following feature set for Windows App SDK:

> -   Supporting raw push for Packaged Win32(MSIX Desktop Bridge)

Reviewer: NIT: space

...for Packaged Win32 (MSIX Desktop Bridge)...

> -   Supporting raw push for Packaged Win32(MSIX Desktop Bridge)

Reviewer: What about MSIX RuntimeBehavior=Win32app?

-   Supporting raw push for Unpackaged Win32.
-   Supporting Visual Cloud Toasts for packaged Win32 apps.
-   Supporting Visual Cloud Toasts for unpackaged Win32 apps.

A Portal Registration flow through
[AAD (Azure Active Directory)](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app)
will also be defined that removes the dependency of the Push Flow with the Partner Center Portal.

> The RemoteIdentifier GUID in this spec maps to the ApplicationId(ClientId) in the AAD App

Author: It shows as "Application (client) ID" in the portal but I will just use ApplicationID.

Registration process. Below is an illustration of the proposed flow through AAD:
![Push Flow For Windows App SDK](Push.png)

Phase 1 (Windows App SDK Preview): WNS Push APIs will only support WIN32 Packaged app (MSIX Desktop
Bridge).

Phase 2 (Windows App SDK V1): WNS Push APIs will support unpackaged Win32 scenarios.

Link to the official Windows App SDK timeline can be found
[here](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md).

# Examples

## In this scenario, the process that Registers the Push Activator and the process specified as the COM server are the same

The code in Main would follow the pattern below:

-   The Registration will take in a CLSID of the COM server as part of the Activator setup. The

>     registration API will simply be a thin wrapper around the BackgroundInfra WinRT APIs and

Author: Removing.

    abstract away the COM details from the developer including registration of the inproc Windows
    App SDK component.

> -   The app will query ApplifeCycle APIs to retrieve an ActivationKind. (Note: This is covered in a
>     seperate AppLifeCycle API spec.)

Reviewer 2: ```suggestion
-   The app will query AppLifecycle APIs to retrieve an ActivationKind. (Note: This is covered in a
    seperate AppLifecycle API spec.)
```

-   If the Activation is Foreground, the app will do 2 things:
    -   It will request a Push Channel Asynchronously with an implementation of Progress and
        Completed event handler.
        -   Expect Channel operations to take around 2-16 minutes in some rare cases (retryable
            errors). In the 90th percentile case, it should be fairly quick operation (in a few
            seconds).
    -   It will subscribe to a In-memory Push event handler hanging off the retrieved Push Channel
        component.
        -   The app should ideally handle the event by setting the Handled property to true to
            prevent Background Activation from launching a new process.
-   If the application is unpackaged, it will use the ProtocolActivator to register the activator
    instead. The app will also need to call into the bootstrapper API to have access to the Windows
    App SDK. Click
    [here](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp)
    for information on how to build and deploy unpackaged applications with Windows App SDK.

```cpp
int main()
{

>     PushNotificationActivationInfo info = nullptr;

Reviewer: NIT: `PushNotificationActivationInfo info{};`

[ES.23: Prefer the {}-initializer syntax](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#es23-prefer-the--initializer-syntax)

Same comment applies throughout

>     bool isComActivationSupported = PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::PushTrigger |

Reviewer: NIT: `bool isComActivationSupported{ Push... };`

                                        PushNotificationRegistrationOptions::ComActivator);
    if (isComActivationSupported)
    {
        // Register the ComActivator guid
        info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::PushTrigger | PushNotificationRegistrationOptions::ComActivator,
        winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));
    }
    else
    {
        info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::ProtocolActivator);
    }

    // Registers the application to receive push notifications
    auto token = PushNotificationManager::RegisterActivator(info);

    // Check to see if the WinMain activation is due to a Push Activator
    auto args = AppInstance::GetCurrent().GetActivatedEventArgs();

>     auto kind = args.Kind();

Reviewer: NIT: `auto kind{ args.Kind() };`

service can triggering an incoming <br/> call toast by Pushing a Payload to the WNS service. Push is
generally used for time critical <br/> events that require immediate user/developer action.

-   [The Push notification WinRT Public APIs](https://docs.microsoft.com/uwp/api/windows.networking.pushnotifications?view=winrt-19041)
    Defines the classes that encapsulate Push notifications. (UWP and Win32)
-   [The Windows Push Notification Services (WNS) Overview](https://docs.microsoft.com/windows/uwp/design/shell/tiles-and-notifications/windows-push-notification-services--wns--overview)
    Defines how an app today can subscribe to Push Notifications and send a payload down to the
    corresponding client.

The below diagram is an illustration of the overall Push Flow that we have today:<br/>
![Legacy Flow For Windows App SDK](Legacy.png)

## The problems today

**Portal Registration**: The entire Registration flow for Push is dependent on registering an app in
the <br/> partner center portal by
[associating an app with the Windows Store](https://docs.microsoft.com/azure/notification-hubs/notification-hubs-windows-store-dotnet-get-started-wns-push-notification).

> As a result, non-store provisioned <br/> apps cannot use this feature.

Reviewer 3: Can you make a pass and see what's going on with all these `<br/>` in the document? Markdown automatically generates appropriate linebreaks. If you want a new paragraph, just use a blank like:

```
instead of <br/> this style
```

```
use this

style.
```

-   Supporting raw push for Unpackaged Win32.
-   Supporting Visual Cloud Toasts for packaged Win32 apps.
-   Supporting Visual Cloud Toasts for unpackaged Win32 apps.

A Portal Registration flow through
[AAD (Azure Active Directory)](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app)
will also be defined that removes the dependency of the Push Flow with <br/> the Partner Center
Portal. The RemoteIdentifier GUID in this spec maps to the ApplicationId(ClientId) in the AAD App
<br/> Registration process. Below is an illustration of the proposed flow through AAD:<br/>
![Push Flow For Windows App SDK](Push.png)

Phase 1 (Windows App SDK Preview):<br/> WNS Push APIs will only support WIN32 Packaged app (MSIX
Desktop Bridge).

Phase 2 (Windows App SDK V1):<br/> WNS Push APIs will support unpackaged Win32 scenarios.

Link to the official Windows App SDK timeline can be found
[here](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md).

# Examples

## In this scenario, the process that Registers the Push Activator and the process specified as the COM server are the same

The code in Main would follow the pattern below:

> -   The Registration will take in a CLSID of the COM server as part of the Activator setup. The

Reviewer 3: (No change required, note on writing style)

Instead of "the bar will foo" consider using "the bar foos".  "Will" suggests indeterminate future, while "bar foos" is a stronger statement that it must be so.  This document, once merged, should be read as "now" rather than "some future point." So someone who comes across this in the future doesn't need to wonder "have they already made this happen, can I use it now?" and instead says "ok, I need my bar to foo in order to accomplish this."

  * It will subscribe to a In-memory Push event handler hanging off the retrieved Push Channel component.
    * The app should ideally handle the event by setting the Handled property to true to prevent Background Activation from <br/>
    launching a new process.
* If the application is unpackaged, it will use the ProtocolActivator to register the activator instead. The app will also need to 
call into the bootstrapper API to have access to the Windows App SDK. Click [here](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp) for information on how to build and 
deploy unpackaged applications with Windows App SDK.
```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;

>     if(PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::PushTrigger | PushNotificationRegistrationOptions::ComActivator))

Reviewer 4: Set this to a local bool and use it everywhere below.
For example,
bool isComActivationSupported = PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::PushTrigger | PushNotificationRegistrationOptions::ComActivator);

  * It will subscribe to a In-memory Push event handler hanging off the retrieved Push Channel component.
    * The app should ideally handle the event by setting the Handled property to true to prevent Background Activation from <br/>
    launching a new process.
* If the application is unpackaged, it will use the ProtocolActivator to register the activator instead. The app will also need to 

> call into the bootstrapper API to have access to the Windows App SDK. Click [here](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp) for information on how to build and 

Reviewer 3: Please word-wrap. Use VSCode's prettier extension which will pull in the .editorconfig for consistency.

> deploy unpackaged applications with Windows App SDK.

Reviewer 3: Needs a newline after "." and before "```"

```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;
    bool isComActivationSupported = PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::PushTrigger | PushNotificationRegistrationOptions::ComActivator);
    if(isComActivationSupported)
    {
        // Register the ComActivator guid

>         info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::PushTrigger | PushNotificationRegistrationOptions::ComActivator,

Reviewer 3: Is there another model instead here, like:

```
auto info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::Default);
```

... where "default" triggers this "is supported" thing, but _inside_ the activation constructor? Or is there a clear reason to make the app developer have this in their application?

Author: Because we support out-of-proc activation for packaged apps, they are allowed to register only a PushTrigger for process A and ComActivator for process B. I think default would be confusing for this case. Also, making the developer know exactly which flag is being used for RegisterActivator will allow them to know which flag to Unregister with.

        winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));
    }
    else
    {
        info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::ProtocolActivator);
    }

    // Registers the application to receive push notifications
    auto token = PushNotificationManager::RegisterActivator(info);

    // Check to see if the WinMain activation is due to a Push Activator

>     auto args = AppInstance::GetCurrent().GetActivatedEventArgs();

Reviewer 4: Thanks for updating this!

    auto kind = args.Kind();

    if (kind == ExtendedActivationKind::Push)
    {

>         PushNotificationReceivedEventArgs pushArgs = args.Data().as<PushNotificationReceivedEventArgs>();

Reviewer 3: Not necessary; the type is already stated in the trailing `.as<T>()`

  * It will subscribe to a In-memory Push event handler hanging off the retrieved Push Channel component.
    * The app should ideally handle the event by setting the Handled property to true to prevent Background Activation from <br/>
    launching a new process.
* If the application is unpackaged, it will use the ProtocolActivator to register the activator instead. The app will also need to 
call into the bootstrapper API to have access to the Windows App SDK. Click [here](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment?tabs=cpp) for information on how to build and 
deploy unpackaged applications with Windows App SDK.

```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;

>     bool isComActivationSupported = PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::PushTrigger | PushNotificationRegistrationOptions::ComActivator);

Reviewer 5: nit: maybe some extra lines for clarity?

Reviewer 5: Same for the other code snippets.

>     if(isComActivationSupported)

Reviewer 5: nit: `if (isComActivationSupported)`

Reviewer 5: Same for the other code snippets.

```cpp
int main()
{

>     PushNotificationRegistrationOptions pushOptions = PushNotificationManager::GetPushActivatorOptions(true);

Reviewer 5: This way, there is no need to look for GetPushActivatorOptions for Process B. We have the confidence to pass a COM flag on Process B.

>     PushNotificationRegistrationOptions pushOptions = PushNotificationManager::GetPushActivatorOptions(true);

Reviewer 4: IDL changes will be needed in the doc. Also please make sure the idl compiles when you change it in the doc.

    }

    // Unregisters the inproc COM Activator before exiting

>     PushNotificationManager::UnregisterActivator(token, pushOptions);

Reviewer 5: If you pass the flags as is and it is ProtocolActivator, we prevent the app to get notifications, which we don't want.

Author: There needs to be a check if we returned Protocol/ComActivator from GetPushActivatorFlags. And unregister accordingly.

        }
    }

    if(isComActivationSupported)
    {
        // Unregister the ComActivator from inproc activation

>         PushNotificationManager::UnregisterActivator(token, PushNotificationRegistrationOptions::ComActivator);

Reviewer 3: Maybe this could be `PushNotificationManager::UnregisterActivator(token)` because the registration step remembers whether it was registered for Com or launch or...

Author: We have to let developer decide which flag they want to unregister. They should call UnregisterActivator with PushTrigger if they no longer want to receive PushNotifications. They should call UnregisterActivator with ComActivator to unregister the calling app from inproc activation.

A common pattern will be RegisterActivator(ComActivator | PushTrigger) -> UnregisterActivator(ComActivator).

Reviewer 3: I see. I'd assumed that a developer would register twice, once per activation kind, rather than or-ing together flags, which meant that they got unique tokens back from registration.

Is there an "unregister all my activators" in case the app can't remember the tokens it got?

Author: We do not have an "unregister all my activators" function. Should we consider adding an API like this?

Reviewer: I'd think so. Handy for apps that do conditional activations (varying options) can have 1 cleanup call `UnregisterActivatorAllKinds()` and not have to track which options they original passed in

Author: We can store the RegistrationTokens for the user and revoke all tokens when they call this function. Should RegisterActivator then not return a token and we just handle it for the user? And UnregisterActivator just accept flags? (What @jonwis said above)
`void RegisterActivator(PushNotificationActivationInfo details);`

`void UnregisterActivator(PushNotificationRegistrationActivators options);

    }

    return 0;
}
```

## In this scenario, the process that Registers the Push Trigger and the process specified as the COM server are different.
This only works for packaged applications that support PushTrigger and ComActivator flags in RegisterActivator. Out-of-proc activation is
not supported for unpackaged apps.

> Process A (Registration of the Push Trigger only):

Reviewer 3: Space needed between paragraphs, so this should probably be:

```
This not only works ... for unpackaged apps.

Process A ... :

``cpp
```

```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;
    if(PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::PushTrigger))
    {
        // Register the ComActivator guid
        info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::PushTrigger,

>         winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));

Reviewer 3: Add indent

    }
    else
    {

>         winrt::throw_hresult(HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED));

Reviewer 4: Should we just return an error code here rather than throw? int == error code

```cpp
int main()
{

>     PushNotificationRegistrationOptions pushOptions = PushNotificationManager::GetPushActivatorKind(false);

Reviewer 6: GetPushActivatorOptions(false);??

Author: This API call will be the equivalent of IsSupported, which will return flags for the developer to call into RegisterActivator. Packaged apps should set this isInProcActivation parameter to true if the calling process will handle background activation.

        }
    }

    if(isComActivationSupported)
    {
        // Unregister the ComActivator from inproc activation
        PushNotificationManager::UnregisterActivator(token, PushNotificationRegistrationOptions::ComActivator);
    }

    return 0;
}
```

## In this scenario, the process that Registers the Push Trigger and the process specified as the COM server are different.

This only works for packaged applications that support PushTrigger and ComActivator flags in

> RegisterActivator. Out-of-proc activation is not supported for unpackaged apps.

Reviewer 2: ```suggestion
RegisterActivator. out-of-process activation is not supported for unpackaged apps.
```

        }
    }

    if(isComActivationSupported)
    {
        // Unregister the ComActivator from inproc activation
        PushNotificationManager::UnregisterActivator(PushNotificationRegistrationActivators::ComActivator);
    }

    return 0;
}
```

## In this scenario, the process that Registers the Push Trigger and the process specified as the COM server are different.

This only works for packaged applications that support PushTrigger and ComActivator flags in
RegisterActivator. Out-of-process activation is not supported for unpackaged apps.

Process A (Registration of the Push Trigger only):

```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;
    if (PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationActivators::PushTrigger))
    {
        // Register the ComActivator guid
        info = PushNotificationActivationInfo(PushNotificationRegistrationActivators::PushTrigger,

>             winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));

Reviewer 3: If this is a sample GUID created by the application, consider hiding it here or making it invalid so people don't cut-and-paste this into their code, like:

```c++
// Be sure to create your own activator ID for each app and registration
static const winrt::guid g_myActivatorId("BACCFA91-F1DE-4CA2-xxxx-90BE66934EC6");

...

info = PushNotificationActivationInfo(PushNotificationRegistrationActivators::PushTrigger, g_myActivatorId);
```

```cpp
int main()
{

>     PushNotificationRegistrationOptions pushOptions = PushNotificationManager::GetPushActivatorKind(true);

Reviewer 6: GetPushActivatorOptions(true);??

```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;
    if(PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::ComActivator))
    {
        // Register the ComActivator guid
        info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::ComActivator,
        winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));
    }
    else
    {

>         winrt::throw_hresult(HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED));

Reviewer 4: Same here.  Just return error code?

Author: Yup. We should return error code.

    return 0;
}
```

Process B (Register the inproc COM server and handle the background activation):

```cpp
int main()
{
    PushNotificationActivationInfo info = nullptr;
    if (PushNotificationManager::IsActivatorSupported(PushNotificationRegistrationOptions::ComActivator))
    {
        // Register the ComActivator guid
        info = PushNotificationActivationInfo(PushNotificationRegistrationOptions::ComActivator,
            winrt::guid("BACCFA91-F1DE-4CA2-B80E-90BE66934EC6"));
    }
    else
    {

>         return ERROR_NOT_SUPPORTED;

Reviewer: Why not E_NOTIMPL? (the typical HRESULT used instead of Win32's ERROR_NOT_SUPPORTED)

  }
}
```
## Registration of the Push Activator for unpackaged apps for protocol activation with PushNotificationsLongRunningTask
The main difference here for unpackaged applications is that RegisterActivator will consume ProtocolActivator</br>
and register with PushNotificationsLongRunningTask, a long running task that is shipped with PRmain, to handle</br>
receiving raw notifications.

**NOTE:** Packaged applications running on Windows builds earlier than 19041 will need to use ProtocolActivator</br>
due to SetTaskEntryPointClsid() being unavailable on those builds.

```cpp
int main()
{
    // Take a dependency on Windows App SDK preview 

>     const UINT32 majorMinorVersion{ 0x00000008 }; 

Author: Yes

>     const UINT32 majorMinorVersion{ 0x00000008 }; 

Reviewer 5: Where the developer should look at to correctly set the majorMinorVersion?

>     const UINT32 majorMinorVersion{ 0x00000008 }; 

Author: I believe they have to get the version themselves. They will run ProjectReunionInstaller and use that installed version

    PCWSTR versionTag{ L"preview" }; 
    const PACKAGE_VERSION minVersion{};

>     const HRESULT hr{ MddBootstrapInitialize(majorMinorVersion, versionTag, minVersion) }; 

Author: If it cannot find a DDLM to connect to it will fail.

    // Check the return code. If there is a failure, display the result.
    if (FAILED(hr)) 
    { 
        wprintf(L"Error 0x%X in MddBootstrapInitialize(0x%08X, %s, %hu.%hu.%hu.%hu)\n", 
            hr, majorMinorVersion, versionTag, minVersion.Major, minVersion.Minor, minVersion.Build, minVersion.Revision); 
        return hr; 
    }

>     PushNotificationRegistrationOptions pushOptions = PushNotificationManager::GetPushActivatorOptions(true);

Reviewer 4: Might want to remove that version and bootstrap stuff in the top for this sample.

Reviewer 4: Unless this is critical for the workings of our API in some way?

Author: All unpkg apps must call MddBootstrapInitialize/MddBootstrapShutdown so I figured our sample code should follow the WindowsAppSDK guidelines.

When an unpkg application exits, it removes its reference from DDLM. So that PRfwk can be serviced when all unpkg references are released.

Author: Also it adds the PRfwk to the packaged graph of the unpkg app which gives them access to our apis.

Reviewer 4: Gotcha...might want to add a comment next to that to say this is critical code for packaged apps to function with the framework

>     PushNotificationRegistrationOptions pushOptions = PushNotificationManager::GetPushActivatorOptions(true);

Reviewer 4: For Unpackaged apps, what does it mean to set inproc to false? Does that mean that we launch a new process when the app is running?

Reviewer 4: It would be nice to have that support. @danielayala94 How could we support out-of-proc activation in your COM design?

    // Unpackaged applications do not need guid to register activator

>     PushNotificationActivationInfo info(

Reviewer 4: Agreed

>     PushNotificationActivationInfo info(

Author: We should also add overload functions for Register/UnregisterActivator. Good idea!

        pushOptions,
        nullptr);

>     // Registers the unpackaged applicaiton with the PushNotificationsLongRunningTask

Author: Probably don't have to have to mention the full impl of unpkg registration. In the appendix it mentions this component and I thought it would be helpful to have a brief comment mentioning where the LRP is being used.

                    {
                        auto payload = args.Payload();

                        // Do stuff to process the raw payload

                        // Stop the subsequent background activation from launching process again with this payload
                        args.Handled(true);
                    });

                        // Persist the channelUri and Expiry in the App Service for subsequent Push operations
                }
                else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
                {
                    LOG_HR_MSG(result.ExtendedError(), "We hit a critical non-retryable error with channel request!");
                }
            });

        // Draw window and other foreground UI stuff here

        // Unregister the Push event for Foreground before exiting
        auto result = channelOperation.GetResults();
        if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
        {
            result.Channel().PushReceived(pushToken);
        }
    }

    // To unregister Push Notifications

>     PushNotificationManager::UnregisterActivator(nullptr, pushOptions);

Reviewer 5: Only if the developer doesn't want to receive notifications anymore, they should call this.

If the developer is expecting to launch an app after receiving a notification, this call should be removed.

Reviewer 4: Correct. UnregisterActivator simple stops PushNotifications. Probably a bad idea to do this in reality.

Author: Right. The comment should probably be, "To completely remove notifications from the application" call this.

Reviewer 4: The problem here is that the Unregister pattern is actually needed to remove the COM registeration. It doesn't mean that Push will stop working in any way. It's kinda like an anti-pattern. Do you have any ideas as to how we can mitigate this?

>     MddBootstrapShutdown(); 

Reviewer 5: What is the purpose of this call?

Reviewer 4: All this bootstrap stuff should be removed from the sample code right?

Reviewer 4: Unless it is critical to the workings of our API in some indirect way?

Author: When an unpkg application exits, it removes its reference from DDLM. So that PRfwk can be serviced when all unpkg references are released.

    };

    [flags]

>     enum PushNotificationRegistrationOptions

Reviewer 4: No need to rename this. Lets keep as is

Author: In the PushNotifications PR, Raymond asked us to change this to *Options so I was just updated the spec with what is in main.

Reviewer 4: Ah got it. Thanks for the context

```

## Manifest Registration

> For MSIX, the COM activator GUID and the exe path need to be registered in the manifest. The launch

Reviewer: NIT: ...and the exe need to be...

One specifies an exe in the manifest (as a relative path/filename) so you don't really specify the path in the manifest (some manifest features do let you specify an actual/full path, but I don't think this is one of them)

> args would need to be pre set to a well-known string that defines Push Triggers.

Reviewer: NIT: pre-set ?

```

# API Details

> ```c# (but really MIDL3)

Reviewer 2: This is not valid syntax, let's not jam up the machines parsing this file.

```suggestion
```c#
```

    };

    [flags]
    enum PushNotificationRegistrationOptions
    {
        PushTrigger = 0x1, // Registers a Push Trigger with Background Infrastructure

>         ComActivator = 0x2, // Registers the Project Reunion Background Task component as an InProc COM server

Reviewer 2: ```suggestion
        ComActivator = 0x2, // Registers the Background Task component as an in-process COM server
```

        ProtocolActivator = 0x4, // Registers an application with the PushNotificationsLongRunningTask to be activated via protocol
    };

    // An abstraction over the activation Registration flow
    runtimeclass PushNotificationActivationInfo
    {
        // Initialize using a Registration option and optionally defined parameters like manifest defined activatorId
        // 1) If kind = PushTrigger is specified, only the Push Trigger will be Registered with Background Infra

>         // 2) If kind = ComActivator is specified, the Project Reunion Background Task component will be Registered as an InProc COM server

Reviewer: the what Background Task component? ;-)

Reviewer 2: ```suggestion
        // 2) If kind = ComActivator is specified, the Background Task component will be Registered as an in-process COM server
```

    };

    [flags]

>     enum PushNotificationRegistrationOptions

Reviewer 3: Consider a name like `PushNotificationActivators` or similar. It's no longer general options about registering activation, it's "which kind of activator does a push notification for this registration use."

    {

>         PushTrigger = 0x1, // Registers a Push Trigger with Background Infrastructure

Author: Yes.

        // Register an activator using an ActivationInfo context and return a RegistrationToken
        static PushNotificationRegistrationToken RegisterActivator(PushNotificationActivationInfo details);

        // Unregister any activator if present using a token and PushNotificationRegistrationOptions
        // 1) If kind = PushTrigger is specified, the trigger itself will be removed
        // 2) If kind = ComActivator is specified, the Project Reunion Background Task component will no longer act as an InProc COM Server
        static void UnregisterActivator(PushNotificationRegistrationToken token, PushNotificationRegistrationOptions options);

        // Unregister the protocol activator registered for the calling application

>         static void UnregisterProtocolActivator();

Reviewer 4: Please remove. Just use the UnregisterActivator function above

        // Register an activator using an ActivationInfo context and return a RegistrationToken
        static PushNotificationRegistrationToken RegisterActivator(PushNotificationActivationInfo details);

        // Unregister any activator if present using a token and PushNotificationRegistrationOptions
        // 1) If kind = PushTrigger is specified, the trigger itself will be removed
        // 2) If kind = ComActivator is specified, the Project Reunion Background Task component will no longer act as an InProc COM Server

>         // 3) If kind = ProtocolActivator is specified, the application will be unregistered from the long running process that handles activation

Reviewer 6: @pmpurifoy should we also explicitly mention token to be nullptr for ProtocolActivator?

    static runtimeclass PushNotificationManager
    {
        // Register an activator using an ActivationInfo context and return a RegistrationToken

>         static void RegisterActivator(PushNotificationActivationInfo details);

Reviewer 4: A little concerned that we removed the token here. The token is being used to infer state like the BackgroundTaskName in the implementation today. How would that flow work without a token?

Author: PushNotificationRegistrationToken moving to private struct and we cache generated tokens by the user to revoke when given unregister flag.

        // Register an activator using an ActivationInfo context and return a RegistrationToken
        static PushNotificationRegistrationToken RegisterActivator(PushNotificationActivationInfo details);

        // Unregister any activator if present using a token and PushNotificationRegistrationOptions
        // 1) If kind = PushTrigger is specified, the trigger itself will be removed

>         // 2) If kind = ComActivator is specified, the Project Reunion Background Task component will no longer act as an InProc COM Server

Reviewer 2: ```suggestion
        // 2) If kind = ComActivator is specified, the Background Task component will no longer act as an InProc COM Server
```

        // 3) If kind = ProtocolActivator is specified, the application will be unregistered from the long running process that handles activation
        //    and the token parameter will be unused so it should be set to nullptr
        static void UnregisterActivator(PushNotificationRegistrationToken token, PushNotificationRegistrationOptions options);

        // Request a Push Channel with an encoded RemoteId from WNS. RemoteId is an AAD identifier GUID
        static Windows.Foundation.IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> CreateChannelAsync(Guid remoteId);

        // Applications will call this to check which flags are supported for RegisterActivator
        static Boolean IsActivatorSupported(PushNotificationRegistrationOptions options);
    };
}

```

# Appendix

> -   For all OS images before 21H1 (19043), the RemoteId will be a noop and the native platform will

Reviewer: "For all Windows releases before..."

Reviewer 2: These are officially documented as _versions_, not _releases_. But otherwise agree with the suggestion.

```suggestion
-   For all versions of Windows 10 prior to Version 21H1 (build 19043), the RemoteId is unused and the native platform will
```

>     send the PFN (PackageFamilyName) in the channel request protocol. The WNS server will maintain

Reviewer: "...send the PackageFamilyName in the channel..."

Remove PFN. Please don't use that acronym as its ambiguous (PackageFamilyName and PackageFullName have the same abbreviation, so we strongly discourage using 'PFN').

Reviewer 2: ```suggestion
    send the Package Family Name in the channel request protocol. The WNS server will maintain
```

    an internal mapping of PFN -> RemoteId and will return a channelUri with the encoded RemoteId
    after querying the map.

> -   For OS images 21H1(19043) and beyond, the Windows App SDK component will call into newly added

Reviewer: NIT: OS images -> Windows releases

Reviewer 2: Would use _versions_, see earlier comments.

>     Closed source APIs in the OS to actually pass a RemoteId in the Channel Request operation.

Reviewer 2: Don't need to call out its open/closed state. For the purposes of this spec, it's an OS provided API. Anything beyond that is out of scope.

Same suggestion applies to other instances below.

```suggestion
    APIs in the OS to pass a RemoteId in the Channel Request operation.
```

-   For unpackaged Win32 apps, the Windows App SDK component will call into Closed source APIs to
    register the unpackaged Win32 process with the RemoteId.
-   A long running Windows App SDK component will be needed to intercept Push payloads from the
    native platform and Launch the corresponding unpackaged Win32 app. We will be using Protocol
    Activation via WinMain to launch the unpackaged Win32 process from the long running service. The
    requirement is mainly because the native client today does not have support to launch Win32 apps
    directly in response to a Push payload unless they are already running in which case we simply

>     marshall the payload back to the process via a registered callback sink.

Reviewer 2: ```suggestion
    marshal the payload back to the process via a registered callback sink.
```

> -   For packaged applications like MSIX Desktop Bridge, we buid a thin wrapper around existing

Reviewer 2: MSIX Desktop Bridge isn't a packaged application. And we've already established the difference between unpackaged/packaged earlier, so it's just repetitive at this point.

```suggestion
-   For packaged applications, we build a thin wrapper around existing
```

        // Unregister any activator if present using a token and registrationKind
        // 1) If kind = PushTrigger is specified, the trigger itself will be removed
        // 2) If kind = ComActivator is specified, the Project Reunion Background Task component will no longer act as an InProc COM Server
        // 3) If kind = ProtocolActivator is specified, the application will be unregistered from the PushNotificationsLongRunningTask

>         static void UnregisterActivator(PushNotificationRegistrationToken token, PushNotificationRegistrationOptions options);

Reviewer 5: If we don't need a token for unpackaged, probably we need an overload of this function.

Just my opinion, not sure what the opinion of Jon or Howard would be on this case.

        // Request a Push Channel with an encoded RemoteId from WNS. RemoteId is an AAD identifier GUID
        static Windows.Foundation.IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> CreateChannelAsync(Guid remoteId);

        // Developers should call this to know with which PushNotificationRegistrationKind to call RegisterActivator. 
        // Packaged applications should set the isInProcActivation to true if they want the current process to handle background activation. 

>         // Unpackaged applications should always set the isInProcActivation to false.

Reviewer 5: Why? Out-of-proc is unsupported for unpackaged apps.

Reviewer 4: Hmm can we support it for unpackaged?
